src/Pos/Generator/BlockEvent.hs

```diff
@@ -155,360 +339,49 @@ newtype Chance = Chance {getChance :: Rational}
-        len <- getRandomR (1, length blockchain)
-        let
-            (blockIndices, remainingBlockchain) = splitAtNewestFirst len $
-                toNewestFirst blockchain
-            -- 'NE.fromList' is valid here because 'len >= 1'.
-            blockIndices' = over _NewestFirst NE.fromList blockIndices
-            remainingBlockchain' = nonEmptyOldestFirst $
-                toOldestFirst remainingBlockchain
-            ev = BlkEvRollback $ BlockEventRollback
-                { _berInput    = blockIndices'
-                , _berOutValid = BlockRollbackSuccess
-                }
-            gbes = maybe GbesStartingState GbesBlockchain remainingBlockchain'
-        return (Just ev, gbes)
-    -- Success without rollback (with apply).
-    genBlockInBlockchain blockchain (IsFailure False) (IsRollback False) = do
-        let tip = NE.last (getOldestFirst blockchain)
-        blockIndices <- genBlockIndices (tip + 1)
-        let
-            ev = BlkEvApply $ BlockEventApply
-                { _beaInput = blockIndices
-                , _beaOutValid = BlockApplySuccess
-                }
-            gbes = GbesBlockchain (blockchain Smg.<> blockIndices)
-        return (Just ev, gbes)
-
-splitAtNewestFirst ::
-    forall a.
-       Int
-    -> NewestFirst NE a
-    -> (NewestFirst [] a, NewestFirst [] a)
-splitAtNewestFirst = coerce (NE.splitAt @a)
-
-nonEmptyOldestFirst ::
-    forall a.
-       OldestFirst [] a
-    -> Maybe (OldestFirst NE a)
-nonEmptyOldestFirst = coerce (nonEmpty @a)
-
-{- NOTE: Reordering corner cases
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-
-When reordering blocks to get an invalid sequence, there's a corner case:
-if there's only one block, it's impossible to get an invalid ordering.
-The reason one might have only one block depends on the means of generation,
-but for 'genBlockIndices' there are two possible causes:
-
-* we rolled 'blockIndexEnd' equal to 'blockIndexStart'
-* 'blockIndexMax' is 1
-
-To avoid dealing with this, we append one more block. The downside is that
-appending this block may cause the sequence to be longer than 'blockIndexMax'.
-
--}
+    checkCount = sum (hhStatusEnd :: HhStatusMap)
+    (hhStatusEnd, revBs') = go Map.empty [] bs
+    bs' = reverse revBs'
+    -- NB. 'go' is tail-recursive.
+    go hhStatusSoFar revNewBs = \case
+        [] -> (hhStatusSoFar, revNewBs)
+        (ev:evs) -> case blkEvTip ev of
+            Nothing -> go hhStatusSoFar (ev:revNewBs) evs
+            Just tipHh -> case Map.lookup tipHh hhStatusSoFar of
+                Nothing ->
+                    let
+                        snapSave = BlkEvSnap (SnapshotSave (hhSnapshotId tipHh))
+                        needSnapSave =
+                            -- We tie the know here to determine whether to actually emit
+                            -- the command to save the snapshot.
+                            Map.findWithDefault 0 tipHh hhStatusEnd > 0
```

Comment:
Yes, we tie the knot, it's 100% intentional and tested. The idea is that we need to save the snapshot only if we later use it, and we know whether we use it only ofter traversing the entire structure. So we use this result "from the future" via laziness, and it doesn't go into a loop because there's no data dependency.